BLAKE3-team/BLAKE3
# Initial TIS-CI configuration

Verifying all test vectors with TIS-CI for four architectures: x86 and ppc, 32-bit and 64-bit.

Feel free to merge it to a test branch (like **tis-ci**) before deciding to put it on master.

About the files that are included:

*   `tis.config` :  
    The main file describing the execution paths to analyze (i.e. your tests) and the analysis options for all the cases.  
    (For more information see: [https://docs.ci.trust-in-soft.com/reference/configuration-file](https://docs.ci.trust-in-soft.com/reference/configuration-file))
*   `c/gen_tis.config.py` :  
    The test script `test.py`, modified to 1) extract all the test vector data and put it into separate files in the directory `tis-ci/test_vectors`, 2) generate the `tis.config` file that mirrors all the tests that are usually ran by `test.py`.  
    (This script is a prototype solution which contains a lot of redundancy with `test.py`. The project's maintainers should have a better idea how to handle this, e.g. leave it as a separate script, incorporate into `test.py`, etc.)
*   `tis-ci/test.c` :  
    A test driver which wraps the main function in order to additionally check whether the test results are correct and whether the main function's return value is zero.  
    (This is optional. Usually TIS-CI only checks if there are no UB on the execution path as checking test results is redundant: the normal test suite is supposed to already handle this. However, it happened to be pretty easy and quick to add the test output checks for BLAKE3, also it proved helpful as it has let us discover some wrong test results on big-endian architectures: [https://github.com/BLAKE3-team/BLAKE3/pull/118](https://github.com/BLAKE3-team/BLAKE3/pull/118))
*   `tis-ci/test_vectors` :  
    The data extracted from test\_vectors/test\_vectors.json to separate files which can be thus be used as input.
*   Also `c/main.c` is slightly modified in order to facilitate the test results checking done by tis-ci/test.c.
*   And the TIS-CI badge was added to `README.md`. It shows the verification status for the last commit on the **master** branch, if you want something else (e.g. last commit on any branch) you'd have to modify it.  
    (For more information see: [https://docs.ci.trust-in-soft.com/reference/status-badge](https://docs.ci.trust-in-soft.com/reference/status-badge))

After merging this PR, you only need to add the repo to TIS-CI from your side. A tutorial is available ([https://docs.ci.trust-in-soft.com/tutorial/set-up-the-continuous-verification#trigger-the-first-analysis](https://docs.ci.trust-in-soft.com/tutorial/set-up-the-continuous-verification#trigger-the-first-analysis)), but it boils down to these steps:

1.  Go to [https://ci.trust-in-soft.com/](https://ci.trust-in-soft.com/) and click on **Sign-up with GitHub**,
2.  During the sign-up process don't forget to grant TIS-CI access to the **BLAKE3-team** organization,
3.  If you already signed up and TIS-CI doesn’t have access to **BLAKE3-team**, you can follow these steps to grant access:  
    [https://docs.ci.trust-in-soft.com/reference/github-organizations#after-sign-up](https://docs.ci.trust-in-soft.com/reference/github-organizations#after-sign-up)
4.  Once in **TIS-CI**, click on **Add Project** (in the top navigation menu),
5.  Choose **BLAKE3-team** from the organization list, then **BLAKE3** from the project list,
6.  Choose the branch where this PR was merged (the one with the `tis.config` file),
7.  Click on **Start verifying**.

The final effect should look similar to this: [https://ci.trust-in-soft.com/projects/jakub-zwolakowski/BLAKE3/6?page=1](https://ci.trust-in-soft.com/projects/jakub-zwolakowski/BLAKE3/6?page=1)

From this point forward all pushes to the repo will automatically trigger a new TIS-CI verification on all the branches which contain a `tis.config` file at the repository root. And you will stay up-to-date on the results thanks to this TIS-CI badge in your README.md.

So after this you should be all good!
